spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/InputBindingLifecycle.java

```diff
@@ -0,0 +1,116 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binding;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.springframework.beans.BeansException;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.SmartLifecycle;
+
+/**
+ * Coordinates binding/unbinding of input channels in accordance to the lifecycle
+ * of the host context.
+ *
+ * @author Marius Bogoevici
+ * @author Ilayaperumal Gopinathan
+ */
+public class InputBindingLifecycle implements SmartLifecycle, ApplicationContextAware {
+
+	private volatile boolean running = false;
+
+	private ConfigurableApplicationContext applicationContext;
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext)
+			throws BeansException {
+		this.applicationContext = (ConfigurableApplicationContext) applicationContext;
+	}
+
+	@Override
+	public void start() {
+		if (!running) {
+			// retrieve the ChannelBindingService lazily, avoiding early initialization
+			try {
+				ChannelBindingService channelBindingService = this.applicationContext
+						.getBean(ChannelBindingService.class);
+				Map<String, Bindable> bindables = this.applicationContext
+						.getBeansOfType(Bindable.class);
+				for (Bindable bindable : bindables.values()) {
+					bindable.bindInputs(channelBindingService);
+				}
+			}
+			catch (BeansException e) {
+				throw new IllegalStateException(
+						"Cannot perform binding, no proper implementation found", e);
+			}
+			this.running = true;
+		}
+	}
+
+	@Override
+	public void stop() {
+		if (running) {
+			try {
+				// retrieve the ChannelBindingService lazily, avoiding early
+				// initialization
+				ChannelBindingService channelBindingService = this.applicationContext
+						.getBean(ChannelBindingService.class);
+				Map<String, Bindable> bindables = this.applicationContext
+						.getBeansOfType(Bindable.class);
+				for (Bindable bindable : bindables.values()) {
+					bindable.unbindInputs(channelBindingService);
+				}
+			}
+			catch (BeansException e) {
+				throw new IllegalStateException(
+						"Cannot perform unbinding, no proper implementation found", e);
+			}
+			this.running = false;
+		}
+	}
+
+	@Override
+	public boolean isRunning() {
+		return running;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return true;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		if (callback != null) {
+			callback.run();
+		}
+	}
+
+	/**
+	 * Return the lowest value to start this bean before any message producing lifecycle
```

Comment:
... highest value to start this bean after any consumer Lifecycle as been started.